backend/src/resolvers/user.resolver.ts

```diff
@@ -8,3 +10,25 @@ export const UserQueryResolver: Pick<QueryResolvers.Type
         return dataSources.databaseAPI.getAllUsers();
     }
 };
+
+export const UserMutationResolver: Pick<MutationResolvers.Type
+    , 'login'
+    | 'signup'
+    > = {
+    // @ts-ignore
+    login: (parent, {email, password}, {dataSources}) => {
+        const user = dataSources.databaseAPI.getUserPerEmail(email);
+        if (!user) return null;
+        if (user.password !== generateSHA512Hash(password)) return null;
+        return createToken(user.id);
+    },
+    // @ts-ignore
+    signup: (parent, {name, email, password}, { dataSources }) => {
+        email = email.toLocaleLowerCase();
+        if (password.length < 8) throw Error('Password is too short!');
+        if (email.indexOf('@') < 1) throw Error('Invalid email address!');
+        if (dataSources.databaseAPI.getUserPerEmail(email) !== undefined) throw Error('Email address already taken!');
+        const userId = dataSources.databaseAPI.createUser(name, email, password);
```

Comment:
I think the check for unique email address validation is missing. so no star for that for now.